File: HOT100/10.回溯/131.分割回文串.cpp
/*
 * @lc app=leetcode.cn id=131 lang=cpp
 *
 * [131] 分割回文串
 */
#include "tools.h"

// @lc code=start
class Solution {
public:
    vector<vector<string>> partition(string s) {
        
        vector<vector<string>> res;
        vector<string> pack;
        int n = s.size();

        auto check = [&](string sub) -> bool
        {
            int l = 0, r = sub.size() - 1;
            while (l <= r)
            {
                if (sub[l] != sub[r]) return 0;
                l ++, r --;
            }
            return 1;
        };

        auto dfs = [&](auto&& self, int sta) -> void
        {
            if (sta == n)
            {
                res.emplace_back(pack);
                return;
            }
            for (int i = sta; i < n; i ++ )
            {
                string sub = s.substr(sta, i - sta + 1);
                if (!check(sub)) continue;
                pack.push_back(sub);
                self(self, i + 1);
                pack.pop_back();
            }
        };

        dfs(dfs, 0);
        return res;
    }
};
// @lc code=end

File: HOT100/HOT100p1/10.2/131.分割回文串.py
#
# @lc app=leetcode.cn id=131 lang=python3
#
# [131] 分割回文串
#
from mytools import *
# @lc code=start
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        n = len(s)
        dp = [[False] * n for _ in range(n)]
        for i in range(n): dp[i][i] = True
        for i in range(n - 1, -1, -1):
            for j in range(i + 1, n):
                if s[i] == s[j]:
                    if j - i >= 2: dp[i][j] = dp[i + 1][j - 1]
                    else: dp[i][j] = True
                else: dp[i][j] = False
        pack = []
        res = []
        def dfs(sta):
            if sta == n:
                res.append(pack[:])
                return
            for i in range(sta, n):
                if not dp[sta][i]: continue
                pack.append(s[sta:i + 1])
                dfs(i + 1)
                pack.pop()
        dfs(0)
        return res
# @lc code=end

